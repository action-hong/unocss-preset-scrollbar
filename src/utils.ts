import {
  escapeSelector as e,
} from 'unocss'

const attributifyRe = /^\[(.+?)(~?=)"(.*)"\]$/

export function toEscapedSelector(raw: string) {
  if (attributifyRe.test(raw))
    return raw.replace(attributifyRe, (_, n, s, i) => `[${e(n)}${s}"${e(i)}"]`)
  return `.${e(raw)}`
}
