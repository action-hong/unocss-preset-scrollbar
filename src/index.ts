import type { Preset } from 'unocss'
import { colorResolver } from '@unocss/preset-mini/utils'
import { toEscapedSelector as e } from 'unocss'
const defaultOption: Required<PresetScrollbarDefaultOption> = {
  scrollbarWidth: '8px',
  scrollbarHeight: '8px',
  scrollbarTrackRadius: '4px',
  scrollbarThumbRadius: '4px',
  scrollbarTrackColor: '#f5f5f5',
  scrollbarThumbColor: '#ddd',
  varPrefix: '',
  numberToUnit: value => `${value / 4}rem`,
}

export interface PresetScrollbarDefaultOption {
  /**
   * scrollbar width
   * @default '8px'
   */
  scrollbarWidth?: string
  /**
   * scrollbar height
   * @default '8px'
   */
  scrollbarHeight?: string
  /**
   * scrollbar track radius
   * @default '4px'
   */
  scrollbarTrackRadius?: string
  /**
   * scrollbar thumb radius
   * @default '4px'
   */
  scrollbarThumbRadius?: string
  /**
   * scrollbar track background color
   * @default '#f5f5f5'
   */
  scrollbarTrackColor?: string
  /**
   * scrollbar thumb background color
   * @default '#ddd'
   */
  scrollbarThumbColor?: string
  /**
   * css variable prefix
   * @default ''
   */
  varPrefix?: string
  /**
   * convert number to unit
   * @default value => `${value / 4}rem`
   * @example
   * numberToUnit: value => `${value / 4}rem`
   * p-4 => padding: 1rem
   * p-4px => padding: 4px
   *
   * @example
   * numberToUnit: value => `${value}rpx`
   * p-4 => padding: 4rpx
   */
  numberToUnit?: (value: number) => string
}

const customRules = {
  'radius': ['track-radius', 'thumb-radius'],
  'w': ['width'],
  'h': ['height'],
  'track-radius': ['track-radius'],
  'thumb-radius': ['thumb-radius'],
}

const numberVarRegex = new RegExp(`^scrollbar-(${Object.keys(customRules).join('|')})-(\\d+?)([a-zA-Z]*?)$`)

export function presetScrollbar(option?: PresetScrollbarDefaultOption): Preset {
  const config = {
    ...defaultOption,
    ...option,
  }

  function resolveVar(name: string) {
    const prefix = config.varPrefix
    return `--${prefix ? `${prefix}-` : ''}scrollbar-${name}`
  }

  return {
    name: 'unocss-preset-scrollbar',
    rules: [
      [/^scrollbar$/, ([_], { rawSelector }) => {
        return `
  ${e(rawSelector)} {
    ${resolveVar('track')}: ${config.scrollbarTrackColor};
    ${resolveVar('thumb')}: ${config.scrollbarThumbColor};
    ${resolveVar('width')}: ${config.scrollbarWidth};
    ${resolveVar('height')}: ${config.scrollbarHeight};
    ${resolveVar('track-radius')}: ${config.scrollbarTrackRadius};
    ${resolveVar('thumb-radius')}: ${config.scrollbarThumbRadius};
    overflow: auto;
    scrollbar-color: var(${resolveVar('thumb')}) var(${resolveVar('track')});
  }
  ${e(rawSelector)}::-webkit-scrollbar-track {
    background: var(${resolveVar('track')});
  }
  ${e(rawSelector)}::-webkit-scrollbar-thumb {
    background: var(${resolveVar('thumb')});
  }
  ${e(rawSelector)}::-webkit-scrollbar {
    width: var(${resolveVar('width')});
    height: var(${resolveVar('height')});
  }
`
      },
      ],
      [/^scrollbar-rounded$/, ([_], { rawSelector }) => {
        return `
  ${e(rawSelector)}::-webkit-scrollbar-track {
    border-radius: var(${resolveVar('track-radius')});
  }
  ${e(rawSelector)}::-webkit-scrollbar-thumb {
    border-radius: var(${resolveVar('thumb-radius')});
  }
`
      }],
      [/^scrollbar-thumb-color-(.+)$/, colorResolver(resolveVar('thumb'), 'scrollbar-thumb')],
      [/^scrollbar-track-color-(.+)$/, colorResolver(resolveVar('track'), 'scrollbar-track')],
      [numberVarRegex, ([_, type, value, unit]) => {
        const val = unit ? value + unit : config.numberToUnit(parseInt(value))
        const vars = customRules[type as keyof typeof customRules]
          .map(v => resolveVar(v))
        return vars.reduce((acc: any, v) => {
          acc[v] = val
          return acc
        }, {})
      }],
    ],
  }
}
