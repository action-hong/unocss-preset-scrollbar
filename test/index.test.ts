import { createGenerator, presetAttributify, presetUno } from 'unocss'
import { describe, expect, it } from 'vitest'
import { presetScrollbar } from '../src'

describe('scrollbar', () => {
  const generator = createGenerator({
    presets: [
      presetUno(),
      presetAttributify(),
      presetScrollbar(),
    ],
  })

  it('scrollbar', async() => {
    const { css } = await generator.generate([
      'scrollbar',
      'scrollbar-rounded',
      'scrollbar-w-4px',
      'scrollbar-radius-2',
      'scrollbar-radius-track-4',
      'scrollbar-radius-thumb-4',
    ].join(' '))

    expect(css).toMatchSnapshot()
  })

  it('scrollbar color', async() => {
    const { css } = await generator.generate([
      'scrollbar-track-color-red',
    ].join(' '))
    expect(css).toMatchSnapshot()
  })

  it('scrollbar custom unit', async() => {
    const { css } = await generator.generate([
      'scrollbar-w-1px',
    ].join(' '))
    expect(css).toMatchSnapshot()
  })

  it('custom value to unit', async() => {
    const generator = createGenerator({
      presets: [
        presetUno(),
        presetScrollbar({
          numberToUnit: value => `${value * 2}px`,
        }),
      ],
    })
    const { css } = await generator.generate([
      'scrollbar-w-1',
      'scrollbar-thumb-radius-2px',
    ].join(' '))
    expect(css).toMatchSnapshot()
  })

  it('should work in atributify mode', async() => {
    const { css } = await generator.generate(`
<div 
  scrollbar="~ rounded w-4px radius-2 radius-track-4 radius-thumb-4">
</div>
`)
    // FIXME: 这里会出现.w-4px{width:4px;}, .rounded{border-radius:0.25rem;} 也很奇怪
    expect(css).toMatchSnapshot()
  })
})
