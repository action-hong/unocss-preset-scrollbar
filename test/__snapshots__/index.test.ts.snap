// Vitest Snapshot v1

exports[`scrollbar > custom value to unit 1`] = `
"/* layer: default */
.scrollbar-thumb-radius-2px{--scrollbar-thumb-radius:2px;}
.scrollbar-w-1{--scrollbar-width:2px;}"
`;

exports[`scrollbar > scrollbar 1`] = `
"/* layer: default */

  .scrollbar {
    --scrollbar-track: #f5f5f5;
    --scrollbar-thumb: #ddd;
    --scrollbar-width: 8px;
    --scrollbar-height: 8px;
    --scrollbar-track-radius: 4px;
    --scrollbar-thumb-radius: 4px;
    overflow: auto;
    scrollbar-color: var(--scrollbar-thumb) var(--scrollbar-track);
  }
  .scrollbar::-webkit-scrollbar-track {
    background: var(--scrollbar-track);
  }
  .scrollbar::-webkit-scrollbar-thumb {
    background: var(--scrollbar-thumb);
  }
  .scrollbar::-webkit-scrollbar {
    width: var(--scrollbar-width);
    height: var(--scrollbar-height);
  }


  .scrollbar-rounded::-webkit-scrollbar-track {
    border-radius: var(--scrollbar-track-radius);
  }
  .scrollbar-rounded::-webkit-scrollbar-thumb {
    border-radius: var(--scrollbar-thumb-radius);
  }

.scrollbar-radius-2{--scrollbar-track-radius:0.5rem;--scrollbar-thumb-radius:0.5rem;}
.scrollbar-w-4px{--scrollbar-width:4px;}"
`;

exports[`scrollbar > scrollbar color 1`] = `
"/* layer: default */
.scrollbar-track-color-red{--un-scrollbar-track-opacity:1;--scrollbar-track:rgba(248,113,113,var(--un-scrollbar-track-opacity));}"
`;

exports[`scrollbar > scrollbar custom unit 1`] = `
"/* layer: default */
.scrollbar-w-1px{--scrollbar-width:1px;}"
`;

exports[`scrollbar > should work in atributify mode 1`] = `
"/* layer: default */
.w-4px{width:4px;}
.rounded{border-radius:0.25rem;}

  .scrollbar {
    --scrollbar-track: #f5f5f5;
    --scrollbar-thumb: #ddd;
    --scrollbar-width: 8px;
    --scrollbar-height: 8px;
    --scrollbar-track-radius: 4px;
    --scrollbar-thumb-radius: 4px;
    overflow: auto;
    scrollbar-color: var(--scrollbar-thumb) var(--scrollbar-track);
  }
  .scrollbar::-webkit-scrollbar-track {
    background: var(--scrollbar-track);
  }
  .scrollbar::-webkit-scrollbar-thumb {
    background: var(--scrollbar-thumb);
  }
  .scrollbar::-webkit-scrollbar {
    width: var(--scrollbar-width);
    height: var(--scrollbar-height);
  }


  [scrollbar~=\\"\\\\~\\"] {
    --scrollbar-track: #f5f5f5;
    --scrollbar-thumb: #ddd;
    --scrollbar-width: 8px;
    --scrollbar-height: 8px;
    --scrollbar-track-radius: 4px;
    --scrollbar-thumb-radius: 4px;
    overflow: auto;
    scrollbar-color: var(--scrollbar-thumb) var(--scrollbar-track);
  }
  [scrollbar~=\\"\\\\~\\"]::-webkit-scrollbar-track {
    background: var(--scrollbar-track);
  }
  [scrollbar~=\\"\\\\~\\"]::-webkit-scrollbar-thumb {
    background: var(--scrollbar-thumb);
  }
  [scrollbar~=\\"\\\\~\\"]::-webkit-scrollbar {
    width: var(--scrollbar-width);
    height: var(--scrollbar-height);
  }


  [scrollbar~=\\"rounded\\"]::-webkit-scrollbar-track {
    border-radius: var(--scrollbar-track-radius);
  }
  [scrollbar~=\\"rounded\\"]::-webkit-scrollbar-thumb {
    border-radius: var(--scrollbar-thumb-radius);
  }

[scrollbar~=\\"radius-2\\"]{--scrollbar-track-radius:0.5rem;--scrollbar-thumb-radius:0.5rem;}
[scrollbar~=\\"w-4px\\"]{--scrollbar-width:4px;}"
`;

exports[`scrollbar > var prefix 1`] = `
"/* layer: default */

  .scrollbar {
    --my-custom-prefix-scrollbar-track: #f5f5f5;
    --my-custom-prefix-scrollbar-thumb: #ddd;
    --my-custom-prefix-scrollbar-width: 8px;
    --my-custom-prefix-scrollbar-height: 8px;
    --my-custom-prefix-scrollbar-track-radius: 4px;
    --my-custom-prefix-scrollbar-thumb-radius: 4px;
    overflow: auto;
    scrollbar-color: var(--my-custom-prefix-scrollbar-thumb) var(--my-custom-prefix-scrollbar-track);
  }
  .scrollbar::-webkit-scrollbar-track {
    background: var(--my-custom-prefix-scrollbar-track);
  }
  .scrollbar::-webkit-scrollbar-thumb {
    background: var(--my-custom-prefix-scrollbar-thumb);
  }
  .scrollbar::-webkit-scrollbar {
    width: var(--my-custom-prefix-scrollbar-width);
    height: var(--my-custom-prefix-scrollbar-height);
  }


  .scrollbar-rounded::-webkit-scrollbar-track {
    border-radius: var(--my-custom-prefix-scrollbar-track-radius);
  }
  .scrollbar-rounded::-webkit-scrollbar-thumb {
    border-radius: var(--my-custom-prefix-scrollbar-thumb-radius);
  }

.scrollbar-thumb-radius-2px{--my-custom-prefix-scrollbar-thumb-radius:2px;}
.scrollbar-w-1{--my-custom-prefix-scrollbar-width:0.25rem;}"
`;
